fix: drop events if above billing limit

kappa90 opened the pull request:
## Problem
We want to drop events in case a user is above their billing limit

## Changes
- If an organization has usage data coming from the billing service, we use it to drop events or recordings, according to their cached limit

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?
- Added tests

kappa90 committed:
fix: drop events if above billing limit

benjackwhite commented on plugin-server/src/worker/ingestion/process-event.ts:
```diff
@@ -73,9 +73,11 @@ export class EventsProcessor {
             if (data['event'] === '$snapshot') {
                 if (team.session_recording_opt_in) {
+                    if (await this.teamManager.isAboveUsageLimit(team.organization_id, 'recordings')) {
+                        return null
```
Wonder if we should temporarily just log something instead of actually dropping for now. That way we can check out the logs and be sure we are not doing anything terrible.

Also wondering if we can add this to statsd or something to track the number of droppages.

The potential negative impact of getting this wrong is very high so I think we should proceed cautiously...

benjackwhite commented on plugin-server/src/utils/db/db.ts:
```diff
@@ -1476,6 +1477,21 @@ export class DB {
         return selectResult.rows[0] ?? null
     }
 
+    public async fetchBillingUsage(organizationId: Team['organization_id']): Promise<BillingUsage | null> {
```
Wondering if this can be done as part of the team load with a JOIN. Saves an extra API request

macobo replied:
Missing unit test.

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -51,6 +52,13 @@ export class TeamManager {
             updateAgeOnGet: false,
         })
+        this.billingUsageCache = new LRU({
+            max: 10000,
```
Problem: If we have more than 10k active orgs at any point then we start doing these queries really rapidly.

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -89,6 +97,42 @@ export class TeamManager {
+        }
+    }
+
+    public async isAboveUsageLimit(organization_id: string, key: string): Promise<boolean> {
```
This needs unit tests to make sure _all_ implicit behavior here is documented (e.g. what happens if postgres query fails)?

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -89,6 +97,42 @@ export class TeamManager {
         }
     }
 
+    public async fetchBillingUsage(organization_id: string): Promise<BillingUsage | null> {
```
Why is this public? :thinking:

macobo commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -30,6 +30,7 @@ const NOT_SYNCED_PROPERTIES = new Set([
 export class TeamManager {
```
We also already have an organizationmanager. Why put organization-level metrics into team manager?

macobo commented on plugin-server/src/worker/ingestion/process-event.ts:
```diff
@@ -73,9 +73,11 @@ export class EventsProcessor {
-
             if (data['event'] === '$snapshot') {
                 if (team.session_recording_opt_in) {
+                    if (await this.teamManager.isAboveUsageLimit(team.organization_id, 'recordings')) {
```
Missing unit tests.

macobo replied:
Missing _unit_ test in plugin-server/tests/worker/ingestion/process-event.test.ts

macobo commented on plugin-server/src/worker/ingestion/process-event.ts:
```diff
@@ -98,6 +100,9 @@ export class EventsProcessor {
                     }
                 }
             } else {
+                if (await this.teamManager.isAboveUsageLimit(team.organization_id, 'events')) {
```
Missing unit tests.

macobo commented on plugin-server/src/types.ts:
```diff
@@ -236,6 +236,7 @@ export enum JobName {
 export type PluginId = Plugin['id']
 export type PluginConfigId = PluginConfig['id']
 export type TeamId = Team['id']
+export type OrgId = Team['organization_id']
```
```suggestion
export type OrganizationId = Team['organization_id']
```

Everything else is fully typed out - let's be consistent.

macobo commented on plugin-server/tests/main/process-event.test.ts:
```diff
@@ -1032,6 +1032,44 @@ test('snapshot event stored as session_recording_event', async () => {
+it('snapshot event not stored if above recordings limit', async () => {
+    await hub.db.postgresQuery('update posthog_team set session_recording_opt_in = $1', [true], 'testRecordings')
+    await hub.db.postgresQuery(
+        `update posthog_organization set usage = '{"recordings": {"usage": 101, "limit": 100}}'`,
```
Aside: I think using jsonb for this was a mistake - more columns would be cheaper and would integrate better with django orm, validations, etc.

macobo commented:
Please take test coverage here super seriously - we need this part of the application to be rock solid as bugs here _will_ mean hard-to-track-down data integrity issues and churned users.

kappa90 committed:
refactor and add tests

kappa90 committed:
fix test

kappa90 commented:
Hey @benjackwhite and @macobo, thanks for your reviews 🙏 

Super noob about this part of the codebase, so forgive me for the mishaps.

I refactored this to use a JOIN in the `fetchTeam` query, so we don't have to cache the billing usage separately.

Regarding using JSONb for the usage, we could refactor it to use integer columns instead, I'm not sure what's the impact on performance but it's needed, I'm up for it.

Regarding testing, I added a unit test about the `isAboveUsageLimit` function (basically the only new one remaining), and I wrote two functional tests in `plugin-server/tests/main/process-event.ts`, but one of them doesn't pass and I have no idea who to fix it (the postgres query hangs).

Also not sure if those tests should stay there or be moved to `plugin-server/tests/worker/ingestion/process-event.ts` as @macobo was suggesting.

I also added logging to statsd, and if we want we can just keep it at that for now, and avoid dropping events. In that case I'd avoid adding the functional tests?

Thank you in advance!

benjackwhite commented on plugin-server/src/worker/ingestion/process-event.ts:
```diff
@@ -73,9 +73,14 @@ export class EventsProcessor {
             if (data['event'] === '$snapshot') {
                 if (team.session_recording_opt_in) {
+                    if (this.teamManager.isAboveUsageLimit(team, 'recordings')) {
+                        this.pluginsServer.statsd?.increment('kafka_queue.single_save.snapshot.above_limit', {
```
Can we key this somehow by "type" as well (so we see if it is "recordings" or "events")

benjackwhite commented on plugin-server/src/worker/ingestion/team-manager.ts:
```diff
@@ -89,6 +89,23 @@ export class TeamManager {
         }
     }
 
+    public isAboveUsageLimit(team: Team, key: string): boolean {
+        const { usage } = team
+        if (!usage) {
+            return false
+        }
+        const usageRecord = usage[key]
+        if (!usageRecord) {
+            return false
+        }
+        const limit = usageRecord.limit
+        if (!limit) {
+            return false
+        }
+        const current = usageRecord.usage
+        return current > limit
+    }
+
```
I'm actually wondering if we want to move this comparison logic into the actual usage record. So we save

```
recordings: {
  usage: 120000,
  limit: 100000,
  rate_limited_until: Date
}
```

This would make the logic here more straight forward and defer the decision to wherever we report the usage, rather than the plugin-server. Not 100% sure if this is the right approach but something like it...

benjackwhite commented on plugin-server/src/worker/ingestion/process-event.ts:
```diff
@@ -73,9 +73,14 @@ export class EventsProcessor {
+                        this.pluginsServer.statsd?.increment('kafka_queue.single_save.snapshot.above_limit', {
+                            team_id: teamId.toString(),
+                        })
+                        return null
```
We should remove the return null whilst we test this.

benjackwhite commented on plugin-server/src/worker/ingestion/process-event.ts:
```diff
@@ -98,6 +103,12 @@ export class EventsProcessor {
+                    this.pluginsServer.statsd?.increment('kafka_queue.single_save.standard.above_limit', {
+                        team_id: teamId.toString(),
+                    })
+                    return null
```
Same here we shouldn't drop until we have tested it in prod

benjackwhite committed:
Added rate limiter code

benjackwhite committed:
Added rate limiter code

benjackwhite committed:
Added todos

benjackwhite committed:
Moved to use DB

kappa90 committed:
refactor to use rate limiter

kappa90 committed:
fix mypy

kappa90 committed:
fix refactor

kappa90 committed:
fix test

kappa90 committed:
fix test

kappa90 committed:
fix test

kappa90 commented:
Splitting this PR into smaller ones.
